readthedocs/core/views/__init__.py

```diff
@@ -53,6 +61,52 @@ def get_context_data(self, **kwargs):
         return context
 
 
+def server_error_404(request, template_name="errors/404/base.html", exception=None):
+    """
+    Serves a 404 error message, handling 404 exception types raised throughout the app.
+
+    Notice that handling of 404 errors happens elsewhere in views and middleware,
+    this view is expected to serve an actual 404 message.
+    """
+
+    context = {}
+    # Properties are set by ProxitoHttp404. We could also have a look at the
+    # subproject_slug
+    if isinstance(exception, ProxitoHttp404):
+        # These attributes are not guaranteed.
+        context.update(
+            {
+                "project": getattr(exception, "project", None),
+                "project_slug": getattr(exception, "project_slug", None),
+                "subproject_slug": getattr(exception, "subproject_slug", None),
+                "version_slug": getattr(exception, "version_slug", None),
+                "language_slug": getattr(exception, "language_slug", None),
+                "path_not_found": getattr(exception, "proxito_path", None),
+            }
```

Comment:
We could probably fill the context in the code below where you check for each type of exception, that woul avoid using `getattr`.

Reply:
It would. But it will also cause a lot of extra code lines.

Reply:
with getattr it is easy to misspell an attribute, and makes it harder to refactor attribute names, having it on each condition is also more clear what exception has what attributes.

Reply:
This is now handled with a pattern that just assumes all additional kwargs for a contextualized Http404 exception as a context variable for its template.